Django Widget Updater

fudgeu opened the pull request:
**This PR has migrations**

Adds the ability to update widgets and re-formats the `widget_metadata` table.

## How updating works

Widget updating is done largely by adding some new entries to a widget's stored metadata:
- `version`, which keeps track of the currently installed version
- `update_method`, which specifies how the widget checks for updates. At the moment, the only value is `github`, which works by checking the latest releases for that widget on Github. In case we or some other third party developer do not want to host widgets on Github (maybe instead using Gitlab or even a custom server), we can implement that specific logic via code and leave it up to the developer to specify which method they want to use for their widget.

When `update_method` is set to `github`, the additional field must also be present:
- `repo`, which is the Github repo that Materia will check for the latest releases available for that widget. This value is in the format of `{owner}/{repo_name}`, much like it would appear in a Github URL. To give an example, the `repo` for Sequencer would be `ucfopen/sequencer-materia-widget`.
  - Additionally, you may also use the github.com URL for the repo itself - e.g. `https://www.github.com/ucfopen/adventure-materia-widget/`, or `github.com/ucfopen/adventure-materia-widget`.

Future update methods may require their own fields, though `repo` will likely be a popular one for similar platforms like Gitlab.

The widget updater determines if a new version is available by comparing the version strings of the currently installed version and the latest returned by the `update_method`. This is done by using a lexicographical spring comparison. Any `v` character at the beginning of a version string is ignored.

## Required changes for widgets

To enable updater support for a widget, a widget's `install.yaml` `meta_data` must define `update_method` and any other fields that `update_method` requires:
- `update_method` - the update method
- for `github`, `repo` must also be set

For example, the following would be added to the Sequencer `install.yaml`:
```yaml
meta_data:
  # ... other metadata fields
  update_method: github
  repo: ucfopen/sequencer-materia-widget
```

Additionally, a new `version.json` file must be present in a widget's compressed `.wigt` file, which should have a 'version' string field. MWDK should auto-generate this file for us based off the `package.json`'s version (but does not at the moment). Lack of this file will default the version to `v1.0.0`.

One thing to note: Because widgets need to define how they update via the `install.yaml`, all currently installed widgets need to be back-filled with this data. Additionally, since no widgets currently exist with these fields in the `install.yaml`, updating a widget to the latest version will empty all these fields from the DB. This will keep occurring until the widget is released with an updated `install.yaml` that includes these fields.

## New commands

A few commands for updating/managing widgets `manage.py widget update <widget_id>` were added:
- `manage.py widget update <widget_id>`
- `manage.py widget update_all`
- `manage.py widget remove <widget_id>`
  - Uninstalls a widget, deleting all instances, permissions, and files.

## New admin UI additions

The widget admin panel also includes new additions to help with widget updating:
- Each widget card now has a 'version' field
- A new 'Update Widgets' section is added, allowing admins to check all widgets for new updates. Admins can then update each widget individually, or all at once.
  - Widgets that were unable to be checked for updates will also appear in their own list, with their errors being output to the browser console. Right now this list will be mostly full of widgets that do not support the updater yet, but can also catch widgets with a broken updater setup or catch update servers that are down.


<img width="779" height="577" alt="Screenshot 2025-08-27 at 5 11 50 PM" src="https://github.com/user-attachments/assets/94b2878a-588c-4777-8063-5b9277c3bc87" />

## Widget metadata table

Lastly, this PR removes the `widget_metadata` table, instead now storing metadata in a new column on the `widget` table. The metadata is stored as plain JSON.

The migration that removes the table and adds the new field (core/0025) will also automatically convert all metadata to and from both formats, forwards and backwards.

fudgeu committed:
Widget updater UI implemented

fudgeu committed:
Reimplement widget update command

FrenjaminBanklin reviewed (commented):
~~It doesn't look like the `/admin/widget` route works, whether from an existing install or a fresh build. Redirects to a 404 immediately every time.~~ This was probably my browser caching files, somehow.

This will be a useful new feature, but I'm not sure I'm on board with adding three new fields to the `Widget` model to facilitate it. We could just as easily store these values as widget metadata like we do `features`, `supported_data`, `demo`, etc. without requiring any model changes.

This does however raise the question of whether we should make any changes to that model at all. The current layout of which properties are attached to `Widget` directly and which are stored as `WidgetMetadata` is a bit arbitrary and this may be a good opportunity to move things around to more sensible locations.

fudgeu committed:
Merge branch 'django-working' into django-widget-updater

fudgeu committed:
Bulk update and new UI

fudgeu committed:
Add metadata conversion to core/0025 migration

fudgeu committed:
Uncomment important line lol

fudgeu committed:
Conversion tool fixes

fudgeu committed:
More fixes to metadata converter

fudgeu committed:
Misc fixes

fudgeu committed:
Merge remote-tracking branch 'origin/django-working' into django-widget-updater

fudgeu committed:
Move migration

fudgeu committed:
Write migration the django-y way

fudgeu commented:
Ok, I've implemented a few rather big changes. I've updated the main PR body to include everything. But, a (not really) brief overview:
- Widget Metadata table no longer exists. Metadata is now stored as a new field on the `widget` table, as JSON. Migration will automatically convert the old table's data to the new format.
- The 3 fields that relate to updating (`version`, `update_method`, and `repo`) are removed from the widget table and moved into the metadata JSON field
- The widget installer now expects to find an auto-generated `version.json` file in a widget's `.wigt` file. MWDK will (but doesn't currently) auto-generate this file at build based off the version found in `package.json`. Lack of this file will default the version to `v1.0.0`
- Added new commands and endpoints for bulk-update checking/updating. See PR body for more details
- Much updated the admin UI to include a new `Update Widgets` section that makes use of the bulk-update-checking. The 'update' field in individual widget cards is no longer there, but the currently widget version remains.
- I've updated how the `github` update method works a bit. It no longer expects a `update_url` field, but instead `repo`. Using just the repo name, we are able to form the required API endpoint ourselves, in one place. Saves a lot of trouble if Github changes things up, and saves widget developers from having to figure out what that endpoint is themselves.
  - On that note, different `update_method`s can define their own required fields now, like how `github` needs a `repo` field. Before, the `update_url` would have been baked into all update methods.

clpetersonucf commented on app/util/widget/installer.py:
```diff
@@ -619,16 +629,123 @@ def install_widget_files(id, clean_name, source_path):
+        if len(releases_json) == 0:
+            return MsgBuilder.failure(msg="Github returned no releases for this widget")
+
+        latest = releases_json[-1]
```
I tested this with Adventure by installing it with the version set to `v2.5.0`, with the latest being `v2.5.1`. It looks like the `releases_json` list is sorted with the newest release first, but this is retrieving the last entry. In the case of Adventure, it was `v2.0.6`. Are there circumstances where we should be checking the last entry instead of the first?

fudgeu replied:
Honestly, not sure how the `releases_json[-1]` thing went under my radar the whole time - I guess it was because I was always manually setting the version to be 'v1.0.0', so latest version never seemed out of place to me, lol. But anyways, that's been fixed.

I've also allowed `repo` to pick up any `github.com` URL, and grab the author and repo names out of that.

clpetersonucf reviewed (changes requested):
This is awesome. Really fantastic work. I added a PR for the MWDK to generate the `version.json` automatically during the build process: https://github.com/ucfopen/Materia-Widget-Dev-Kit/pull/140

The only discrepancy I noticed was the `releases_json` list and referencing the earliest entry instead of the latest. More on that below.

Otherwise, my only thought is to potentially accommodate variants of the `repo` string in the `install.yaml`: could we support a resolved URL (like `github.com/ucfopen/adventure-materia-widget` or `https://github.com/ucfopen/adventure-materia-widget` in addition to `ucfopen/adventure-materia-widget`?

fudgeu committed:
Use latest version, accept full URLs for repo

clpetersonucf committed:
Merge branch 'django-working' into django-widget-updater

clpetersonucf committed:
Repositions migrations since 0027 was already taken

clpetersonucf reviewed (approved):
Updated with `django-working` and adjusted the new migration number. Should be good to go.
